### PR TITLE
Add support power palette order

### PR DESCRIPTION
**Pull request by dragunoff**
As suggested in #16383 and discussed on discord around the same time. This is very similar to `BuildPaletteOrder`. The order is as follows:

Index | RA | TD | d2k
---: | :--- | :--- | :---
10 | Spy Plane | Airstrike | Airstrike
20 | Paradrop | Ion Canon | Fremen
30 | Parabombs | Nuke | Saboteur
40 | Iron Curtain | | Nuke
50 | Chronoshift | | 
60 | Adv. Chronoshift | | 
70 | Nuke | | 
80 | Sonar Pulse | | 
90 | GPS | | 

I have not ordered the TS powers because not all have been implemented yet.

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/SupportPowersWidget.cs [hunk 1]
```diff
@@ -121,7 +121,7 @@ public class SupportPowerIcon
 		public void RefreshIcons()
 		{
 			icons = new Dictionary<Rectangle, SupportPowerIcon>();
-			var powers = spm.Powers.Values.Where(p => !p.Disabled);
+			var powers = spm.Powers.Values.Where(p => !p.Disabled).OrderBy(p => p.Info.SupportPowerPaletteOrder);
```
Comment (teinarss) [inline OpenRA.Mods.Common/Widgets/SupportPowersWidget.cs]:
```
			var powers = spm.Powers.Values.Where(p => !p.Disabled).OrderBy(p => p.Info.SupportPowerPaletteOrder);
```

This should prolly be added to this also https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Widgets/ObserverSupportPowerIconsWidget.cs#L104

**Review by teinarss — CHANGES_REQUESTED** [review]

**Review by abcdefg30 — COMMENTED** [review]
Is there a specific reason you didn't apply this to TS?

**Comment by pchote** [comment]
> I have not ordered the TS powers because not all have been implemented yet.

**Comment by abcdefg30** [comment]
Derp, I can't read any more apparently.

**Comment by dragunoff** [comment]
Rebased and updated with requested changes.

**Review by pchote — COMMENTED** [review]
Minor code style nit, otherwise LGTM.

**Comment by dragunoff** [comment]
Rebased and addressed style nits.

**Review by ghost — APPROVED** [review]
LGTM.

**Comment by Orb370** [comment]
The original reason I brought this up is because Chronoshift/Iron Curtain are time sensitive superpowers but often their hotkeys are different depending on what faction you are, and inconsistent depending on unlock order.

I think both superpowers should be *first*, that way no matter what other superpowers you have at least your hotkey will stay consistent.

**Comment by dragunoff** [comment]
This is what I'm proposing here (allies, soviets, double tech):
изображение

@Orb370 If I understand you correctly you are suggesting that IC and chrono be at the top because they are the most time sensitive powers and should always be available via a hotkey (as there are only 4 hotkeys)?

**Comment by Orb370** [comment]
Yes. But also for people who play different factions, IC/Chrono will be on 3 different hotkeys depending on the faction. Ukraine, Russia, and Any Allied Faction respectively.

If it's always at the top then it doesn't matter what faction you are playing.

**Comment by matjaeck** [comment]
> Yes. But also for people who play different factions, IC/Chrono will be on 3 different hotkeys depending on the faction. Ukraine, Russia, and Any Allied Faction respectively.
> 
> If it's always at the top then it doesn't matter what faction you are playing.

Not quite. The order of the powers does not directly correlate with the hotkey slots: If you only have one power, that power will always be in slot one. The support power slot hotkey always relates to the actual available powers, and only then to their order. So it is very possible that the hotkey for your powers change during the game (i.e. building an airfield before the nuke).


Would it make sense to start with the strongest power, something like this?

Index | RA | 
-- | -- |
10 | Nuke |
20 | Iron Curtain |
30 | Adv. Chronoshift |  
40 | Chronoshift |
50 | Parabombs |
60 | Paradrop | 
70 | Spy Plane |
80 | Sonar Pulse | 
90 | GPS |

**Comment by Orb370** [comment]
That is why I suggest moving them to the front. That way, it always slots into #1, regardless of what support powers you have.

Nukes have a long cooldown, take a long time to reach their target, and are a fire and forget. I think the original ordering is fine as far as that is concerned. I don't think I'd ever lose/win a game if I delayed my nuke by several seconds, but a IC/Chronoshift delayed by even a fraction of a second can make the difference.

**Review by ghost — CHANGES_REQUESTED** [review]
I think @Orb370 has a point here. When we move the nuke and ic/chrono up they will always be on the same hotkeys once they are build, which does make things easier.

**Comment by pchote** [comment]
:+1: to putting IC/Chrono first, then nuke, then everything else.

**Comment by dragunoff** [comment]
I agree with @Orb370. IC/Chrono usage is very time senstive and having a consistent hotkey (except in doble tech scenarios) can help with that. So  something like this: 

Index | RA
-- | --
10 | Iron Curtain
20 | Chronoshift
30 | Adv. Chronoshift
40 | Parabombs
50 | Paradrop
60 | Spy Plane
70 | Nuke
80 | Sonar Pulse
90 | GPS

Which results in this in-game (doble tech, allies, soviet):
изображение

* Getting both chronoshift and adv. chronoshift would be a very rare case so the second chronoshift here can be mostly igrnored.
* Nuke is near the bottom because games often don't reach that stage and if it had a higher priority it would disrupt the order when it becomes available. And as mentioned already - it has a long cooldown and usually the player has enough time to choose the target w/out a hotkey. Up for debate of course.

**Comment by matjaeck** [comment]
To me it would make more sense to have the nuke / ic / chrono all in the first slots (IMO ordered by strenth and have the nuke in slot 1) but you got a point with

> if it had a higher priority it would disrupt the order when it becomes available

So I'm fine with both ways.

**Comment by dragunoff** [comment]
Updated with the new order for RA as discussed.

**Review by ghost — APPROVED** [review]
Looks good with IC at start and nuke at bottom.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f349a3644cec739bc18822b0544f7497cc88c976...222ff88239d56858be6b2329eb6f7a287e9296b1)

**Pull request merged**